### PR TITLE
Migrate to Tebako packager

### PR DESCRIPTION
**Pull request by ronaldtse**
### Metanorma PR checklist

 - [ ] Breaking changes (list related PRs)
 - [ ] Documentation update required ([create task for this](https://github.com/metanorma/metanorma.org/issues/new))
 - [ ] External dependency introduced ([documentation update need](https://github.com/metanorma/metanorma.org/issues/new))
 - [ ] Gem with native library introduced

### CODE REVIEW

**Comment by ronaldtse** [comment]
@maxirmx here's the real error from https://github.com/metanorma/packed-mn/runs/4419965948?check_suite_focus=true:

```
2021-12-05T01:23:57.8451319Z [info]: Compiling /home/runner/work/packed-mn/packed-mn/iso-19156/sources/iso-19156.adoc ...
2021-12-05T01:24:04.2376054Z [relaton] Info: detecting backends:
2021-12-05T01:24:05.1042651Z Fatal Error: No such file or directory @ rb_sysopen - /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd/../../../VERSION
2021-12-05T01:24:05.1193801Z /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd/version.rb:3:in `read': No such file or directory @ rb_sysopen - /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd/../../../VERSION (Errno::ENOENT)
2021-12-05T01:24:05.1196934Z 	from /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd/version.rb:3:in `<module:VERSION>'
2021-12-05T01:24:05.1198286Z 	from /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd/version.rb:1:in `<top (required)>'
2021-12-05T01:24:05.1199546Z 	from /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd.rb:14:in `require'
2021-12-05T01:24:05.1201045Z 	from /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd.rb:14:in `block in <top (required)>'
2021-12-05T01:24:05.1204025Z 	from /__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd.rb:14:in `glob'
```

**Comment by maxirmx** [comment]
Thank you, @ronaldtse !
```
/__tebako_memfs__/lib/ruby/gems/2.7.0/gems/rdf-xsd-3.1.1/lib/rdf/xsd/../../../VERSION
```
I can see that the file is in memfs but ruby cannot find it so it shall be an issue with libdwarfs-wr. I think I would rather test it bottom up:
https://github.com/tamatebako/libdwarfs/issues/51

**Comment by maxirmx** [comment]
Also this is the usual cost payed by the people who try to cut the corners.
I should do unit tests as planned (https://github.com/tamatebako/tebako/issues/19) but I have skipped it.

**Comment by ronaldtse** [comment]
> Also this is the usual cost payed by the people who try to cut the corners. I should do unit tests as planned ([tamatebako/tebako#19](https://github.com/tamatebako/tebako/issues/19)) but I have skipped it.

Usually it's better to pay the price early on rather than run into problems later. Agree 😉

**Comment by ronaldtse** [comment]
Strange to see usage of so much swap... how are we loading the DwarFS image in the binary?

**Comment by maxirmx** [comment]
> Strange to see usage of so much swap... how are we loading the DwarFS image in the binary?

It happens prior to loading image to the binary.  Memory is required for mkdwarfs to create compressed image.  Libzstd

We are loading by INCBIN but it is not related

**Comment by ronaldtse** [comment]
> It happens prior to loading image to the binary. Memory is required for mkdwarfs to create compressed image. Libzstd

Isn't it strange to require 25GB of swap for an image of few hundred MBs...?

**Comment by maxirmx** [comment]
Actually it worked with 17 gb
Pls note ruby-packer workflow has 15 gb setup

**Comment by maxirmx** [comment]
https://github.com/mhx/dwarfs/blob/main/doc/mkdwarfs.md

Note memory limit option.

**Comment by maxirmx** [comment]
@ronaldtse @CAMOBAP @opoudjis 
Not sure what to do with it next but we can definetely talk about it.
I am afraid you would expect to see single executable like ruby-packer and not a build script.  However, my approach has one advantage - if environment is set up packaging takes 5 minutes comparing to ruby-packer's 1 hour

Thank you

**Comment by CAMOBAP** [comment]
> I am afraid you would expect to see single executable like ruby-packer and not a build script

Not a problem from my point of view

@maxirmx I have two questions sofar:
1. Currently all setup logic is placed into the GHA workflow, which is mostly ok, but I think sometimes we will need to troubleshoot some issues locally so it would be nice to convert it to some sh script, how do you think, is it make sense?
2. As far as I see currently we support only linux, right? What is the status of macos and windows?

**Comment by maxirmx** [comment]
> 2\. As far as I see currently we support only linux, right? What is the status of macos and windows?

I have preliminary estimates only.  I think I will be able to have MacOS version of the same functionality in approx 2 weeks. Windows ~ 6 weeks, but it is very prilimiary and shaky estimate.

I did not start either of them yet.

**Comment by maxirmx** [comment]
@ronaldtse 

> Also need documentation for tebako 
here is a brief document here:
https://github.com/tamatebako/tebako/blob/master/doc/DEV-NOTES.md
I encourage everybody ask additional questions or otherwise critisize it.

>Can I confirm that a binary built on Ubuntu 20.04 will work on a vanilla Ubuntu 18.04?

No, the status today is as follows:

- A binary built on Ubuntu 18.04 will work on Ubuntu 20.04.  It is tested now but I use GHA images. I can probably deploy  clean Ubuntu container to make it "more vanilla".

- However,  a  binary built on Ubuntu 20.04 WILL NOT WORK on Ubuntu 18.04  since I am linking shared glibc as ruby-packer does.  There are some options to make 20.04 image work on 18.04.

https://github.com/wheybags/glibc_version_header
https://github.com/crosstool-ng/crosstool-ng
and an overview here:  https://stackoverflow.com/questions/2856438/how-can-i-link-to-a-specific-glibc-version

**Comment by ronaldtse** [comment]
Not sure why macOS build is failing here:

```
/__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/ffi-1.15.4/lib/ffi/library.rb:145:in `block in ffi_lib': Could not open library '/__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/ext/x86_64-darwin/liblibmspack.bundle': dlopen(/__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/ext/x86_64-darwin/liblibmspack.bundle, 5): image not found (LoadError)
8
	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/ffi-1.15.4/lib/ffi/library.rb:99:in `map'
9
	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/ffi-1.15.4/lib/ffi/library.rb:99:in `ffi_lib'
10
	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/lib/libmspack.rb:18:in `<module:LibMsPack>'
11
	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/lib/libmspack.rb:16:in `<top (required)>'
12
	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/excavate-0.3.0/lib/excavate/extractors/cab_extractor.rb:1:in `require'
13
	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/excavate-0.3.0/lib/excavate/extractors/cab_extractor.rb:1:in `<top (required)>'
14
```

But hopefully we are able to migrate macOS to Tebako soon.

**Comment by CAMOBAP** [comment]
@ronaldtse `liblibmspack.bundle` - looks like wrong name it should be any of `mspack.bundle` or `libmspack.bundle`

**Comment by ronaldtse** [comment]
Good find. I wonder why that is the case... seems that the name got an additional prefix.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -4,64 +4,187 @@ on:
...
+  DEPS: deps
+# Versions
+  TEBAKO_TAG: "v0.2.3"
+  RUBY_VER:   "2.7.4"
+  RUBY_HASH:  "3043099089608859fc8cce7f9fdccaa1f53a462457e3838ec3b25a7d609fbc5b"
+  BUNDLER_VER: "2.2.3"
+
 jobs:
+  setup:
+    runs-on: ubuntu-18.04
```
[the remark this answers is not in the document]
Comment (ronaldtse) [reply .github/workflows/linux.yml:39]: This will be handled here: https://github.com/tamatebako/tebako/issues/53

**Comment by ronaldtse** [comment]
The only remaining issue now is the macOS build right now failing due to `liblibmspack.bundle` not found (when running `metanorma help`).

```
2021-12-18T08:48:44.4412450Z /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/ffi-1.15.4/lib/ffi/library.rb:145:in `block in ffi_lib': Could not open library '/__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/ext/x86_64-darwin/liblibmspack.bundle': dlopen(/__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/ext/x86_64-darwin/liblibmspack.bundle, 5): image not found (LoadError)
2021-12-18T08:48:44.4415840Z 	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/ffi-1.15.4/lib/ffi/library.rb:99:in `map'
2021-12-18T08:48:44.4417830Z 	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/ffi-1.15.4/lib/ffi/library.rb:99:in `ffi_lib'
2021-12-18T08:48:44.4419920Z 	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/lib/libmspack.rb:18:in `<module:LibMsPack>'
2021-12-18T08:48:44.4422250Z 	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/lib/libmspack.rb:16:in `<top (required)>'
2021-12-18T08:48:44.4424580Z 	from /__ruby_packer_memfs__/local/vendor/bundle/ruby/2.6.0/gems/excavate-0.3.0/lib/excavate/extractors/cab_extractor.rb:1:in `require'
```

However, lo and behold, I do have this exact path in one of my installs!!

```
PATH: vendor/bundle/ruby/2.6.0/gems/libmspack-0.10.1/ext/x86_64-linux/
```
Screenshot 2021-12-18 at 10 05 49 PM

That root path has a `.bundle/config` file that has:
```sh
$ cat .bundle/config
---
BUNDLE_PATH: "vendor/bundle"
BUNDLE_CACHE_ALL: "true"
```

**Comment by ronaldtse** [comment]
> The only remaining issue now is the macOS build right now failing due to `liblibmspack.bundle` not found (when running `metanorma help`).

This is due to this code where it tries to add the `lib` prefix:
https://github.com/ffi/ffi/blob/7c70b829ab422dceeaf472436f3a38fca6ef8f0b/lib/ffi/library.rb#L44-L55

```ruby
  def self.map_library_name(lib)
    # Mangle the library name to reflect the native library naming conventions
    lib = Library::LIBC if lib == 'c'

    if lib && File.basename(lib) == lib
      lib = Platform::LIBPREFIX + lib unless lib =~ /^#{Platform::LIBPREFIX}/
      r = Platform::IS_WINDOWS || Platform::IS_MAC ? "\\.#{Platform::LIBSUFFIX}$" : "\\.so($|\\.[1234567890]+)"
      lib += ".#{Platform::LIBSUFFIX}" unless lib =~ /#{r}/
    end

    lib
  end
```

And this is where the search happens:
https://github.com/ffi/ffi/blob/7c70b829ab422dceeaf472436f3a38fca6ef8f0b/lib/ffi/library.rb#L95-L104
```ruby
    def ffi_lib(*names)
      raise LoadError.new("library names list must not be empty") if names.empty?

      lib_flags = defined?(@ffi_lib_flags) ? @ffi_lib_flags : FFI::DynamicLibrary::RTLD_LAZY | FFI::DynamicLibrary::RTLD_LOCAL
      ffi_libs = names.map do |name|

        if name == FFI::CURRENT_PROCESS
          FFI::DynamicLibrary.open(nil, FFI::DynamicLibrary::RTLD_LAZY | FFI::DynamicLibrary::RTLD_LOCAL)

        else
```

It's possible that the `libmspack` gem uses the `ffi_lib` method incorrectly.

**Comment by maxirmx** [comment]
@alexeymorozov  kindly pointed to the issue with different memfs mount points 
In Linux we now need  /__tebako_memfs__,   MacOS - /__ruby_packer_memfs__

**Comment by ronaldtse** [comment]
@maxirmx actually we do not need `__ruby_packer_memfs__`. As long as there is one consistent mount point we can accommodate in the code. But of course if the mount point is configurable it would be better.

**Comment by ronaldtse** [comment]
@maxirmx could you help resolve the merge conflicts and merge this PR? Thanks!

**Comment by maxirmx** [comment]
> @maxirmx could you help resolve the merge conflicts ...
Done

**Comment by ronaldtse** [comment]
Thanks @maxirmx ! 

Now, technically, the three remaining failures are due to https://github.com/metanorma/isodoc/issues/367. This is a bug in the metanorma-iso gem, that requires the “sassc” gem. 

However, it would be useful if we can package the “sassc” gem as well so that users can use custom SASS/SCSS files. Let me make that a separate issue.

**Pull request merged**